src/core/Global.ttslua

```diff
@@ -456,358 +268,281 @@ function drawChaostoken(params)
-    merge_distance = 5.0,
-    stackable = true,
+    merge_distance = 5,
+    stackable = true
   })
-  token.use_snap_points=false
+  token.use_snap_points = false
   token.scale(tokenData['scale'])
   return token
 end
 
-function round(params) -- Parameter (int number, int numberDecimalPlaces)
-  return tonumber(string.format("%." .. (params[2] or 0) .. "f", params[1]))
-end
+---------------------------------------------------------
+-- chaos token stat tracker
+---------------------------------------------------------
 
-function roundposition(params) -- Parameter (Table position)
-  return {round({params[1], 2}),round({params[2], 2}),round({params[3], 2})}
+function take_callback(object_spawned, mat)
+  local player = mat.getGUID()
+  local image = object_spawned.getCustomObject().image
+  PULLS[image]                = (PULLS[image] or 0) + 1
+  PLAYER_PULLS[player][image] = (PLAYER_PULLS[player][image] or 0) + 1
 end
 
-function isEqual(params) --Parameter (Table table1, Table table2) returns true if the tables are equal
-  if params[1][1] == params[2][1] and params[1][2] == params[2][2] and params[1][3] == params[2][3] then
-    return true
+function printOrResetStats(_, _, isRightClick)
+  if isRightClick then
+    for key, _ in pairs(PULLS) do
+      PULLS[key] = 0
+    end
+    for playerKey, _ in pairs(PLAYER_PULLS) do
+      for key, value in pairs(PULLS) do
+        PLAYER_PULLS[playerKey][key] = value
+      end
+    end
   else
-    return false
+    local squidKing = "Nobody"
+    printToAll("Overall Game stats")
+    printToAll("------------------------------")
+    printNonZeroTokenPairs(PULLS)
+    printToAll("Individual Stats")
+    printToAll("------------------------------")
+    for playerMatGuid, _ in pairs(PLAYER_PULLS) do
+      local playerColour = MAT_GUID_TO_COLOUR[playerMatGuid]
+      local playerSquidCount = PLAYER_PULLS[playerMatGuid]["https://i.imgur.com/lns4fhz.png"] or 0
+      local playerName = playerColour
+      if Player[playerColour].seated then
+        playerName = Player[playerColour].steam_name
+      end
+
+      printToAll(playerName ..  " Stats", playerColour)
+      printNonZeroTokenPairs(PLAYER_PULLS[playerMatGuid])
+      
+      if playerSquidCount > maxSquid then
+        squidKing = playerName
+        maxSquid = playerSquidCount
+      end
+    end
+    printToAll(squidKing .. " is an auto-fail magnet.", {255, 0, 0})
   end
 end
 
-function isFaceup(params) --Object object
-  if params.getRotation()[3] > -5 and params.getRotation()[3] < 5 then
-    return true
-  else
-    return false
+function printNonZeroTokenPairs(theTable)
+  for key, value in pairs(theTable) do
+    if value ~= 0 then printToAll(IMAGE_TOKEN_MAP[key] .. ': ' .. tostring(value)) end
   end
 end
 
---Difficulty selector script
+---------------------------------------------------------
+-- Difficulty selector script
+---------------------------------------------------------
 
 function createSetupButtons(args)
```

Comment:
Setup is very generic, maybe createBagSetupButtons?

Also good practice to document expected values for anything that just takes the single table parameter style for args

Reply:
this function is called by all the difficulty selectors contained in all scenario setup bags, will not change that for now